mgr/dashboard: Cluster Creation Add multiple hosts at once

aaSharma14 opened the pull request:
Add multiple hosts at once in cluster creation wizard

Fixes: https://tracker.ceph.com/issues/51991
Signed-off-by: Aashish Sharma <aasharma@redhat.com>



<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

aaSharma14 commented:
jenkins test dashboard cephadm

avanthakkar commented:
jenkins test dashboard cephadm

avanthakkar commented:
jenkins test dashboard

avanthakkar commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -57,50 +66,99 @@ export class HostFormComponent extends CdForm implements OnInit {
+        );
+        const indexes = hostnameRange.split('-');
+        let hostNumber: string;
+        for (let i = indexes[0]; i <= indexes[indexes.length - 1]; i++) {
```
`forEach` would be better here IMO. BTW can this whole logic (from `109-137`) moved to a seperate function which just return u final hostnames list..  maybe a `hostnameValidator()` ?

avanthakkar commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -57,50 +66,99 @@ export class HostFormComponent extends CdForm implements OnInit {
+          task: new FinishedTask('host/' + URLVerbs.ADD, {
+            hostname: hostnames
+          }),
+          call: this.hostService.create(hostnames, this.addr, this.allLabels, this.status)
```
So it means for 100 hostnames list 100 API request!! ?🙄 Should we make some changes in backend to instead take list of hostnames and forward request synchronously to cephadm `orch host add` ? @alfonsomthd @epuertat

avanthakkar replied:
Also it'll take "years" to get inventory, facts, etc as it'll refresh `@forallhosts()` in cephadm `serve.py` for 10/100s of hosts.

alfonsomthd replied:
> So it means for 100 hostnames list 100 API requests!! ? Should we make some changes in backend to instead take list of hostnames and forward request synchronously to cephadm `orch host add` ? @alfonsomthd @epuertat

I think so. Maybe we should create a `bulk` option/endpoint to handle this at the backend level. Does
orchestrator support bulk host addition?

pereman2 replied:
> Does orchestrator support bulk host addition?

I think not.

pereman2 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -57,50 +66,99 @@ export class HostFormComponent extends CdForm implements OnInit {
+    const hostnameArray = [];
     const hostname = this.hostForm.get('hostname').value;
+    if (this.hostForm.get('pattern').value === true) {
+      if (hostname.includes('[') && hostname.includes(']')) {
+        const hostnameRange = hostname.replace(/(^.*\[|\].*$)/g, '');
+        const patternBrackets = hostname.substring(
+          hostname.indexOf('['),
```
I think extracting this `if` statement or adding a comment would help.

epuertat reviewed (changes requested):
Good work, @aaSharma14 !

I'd like to see some unit testing for the pattern expansion method.

nizamial09 reviewed (changes requested):
Just tested this and noticed somethings.

1. If I enter a pattern without ticking the `Hostname Pattern` checkbox, the input field is giving me a validation error but not giving me any indication of what is the actual error.
2. I entered the pattern at first and then clicked on the `Hostname Pattern` checkbox and it immediately cleared what I typed on the Hostname field.
3. I put in the pattern like this `ceph-node-[01-02].cephlab.com`, it added the ceph-node-02, but instead of adding ceph-node-01 it started adding ceph-node-001. 
4. Also the wrong validator here doesn't say what's wrong with the field.


Apart from these, some of the observations I have is:
1. We should rename the `Hostname` input field to something like `Hostname/Hostname Pattern` or just `Hostname/Pattern`, otherwise its not specific to where to type the hostname.
2. We should remove the placeholder or at least change it to take in account of the pattern. Currently, its just `mon-123`.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/cypress/integration/orchestrator/workflow/02-create-cluster-add-host.e2e-spec.ts:
```diff
@@ -5,7 +5,8 @@ describe('Create cluster add host page', () => {
     'ceph-node-00.cephlab.com',
     'ceph-node-01.cephlab.com',
-    'ceph-node-02.cephlab.com'
+    'ceph-node-02.cephlab.com',
+    'ceph-node-[3-4].cephlab.com'
```
for me its doing `ceph-node-03.cephlab.com,ceph-node-04.cephlab.com`

aaSharma14 replied:
Since we are using brace-expansion package now, `[3-4]` will expand to `3,4` and `[03-04]` will expand to `03,04`

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.html:
```diff
@@ -35,8 +36,26 @@
+                       for="pattern"
+                       i18n>Hostname Pattern</label>
+                <cd-helper>
+                  <span i18n>Add multiple hosts at once. You can provide a pattern in this format - ceph-node-[00-10].cephlab.com or separated by commas such as ceph-node.cephlab.com_0,ceph-node.cephlab.com_1</span>
```
Done. Thanks

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -57,50 +66,99 @@ export class HostFormComponent extends CdForm implements OnInit {
     const hostname = this.hostForm.get('hostname').value;
+    if (this.hostForm.get('pattern').value === true) {
+      if (hostname.includes('[') && hostname.includes(']')) {
+        const hostnameRange = hostname.replace(/(^.*\[|\].*$)/g, '');
```
The latest change supports all these patterns that you have mentioned.

aaSharma14 commented:
> Just tested this and noticed somethings.
> 
> 1. If I enter a pattern without ticking the `Hostname Pattern` checkbox, the input field is giving me a validation error but not giving me any indication of what is the actual error.
> 2. I entered the pattern at first and then clicked on the `Hostname Pattern` checkbox and it immediately cleared what I typed on the Hostname field.
> 3. I put in the pattern like this `ceph-node-[01-02].cephlab.com`, it added the ceph-node-02, but instead of adding ceph-node-01 it started adding ceph-node-001.
> 4. Also the wrong validator here doesn't say what's wrong with the field.
>    ![wrong_validation](https://user-images.githubusercontent.com/71764184/135035280-f448efec-9e00-4c8c-98e2-3967983e5b6f.png)
> 
> Apart from these, some of the observations I have is:
> 
> 1. We should rename the `Hostname` input field to something like `Hostname/Hostname Pattern` or just `Hostname/Pattern`, otherwise its not specific to where to type the hostname pattern.
> 2. We should remove the placeholder or at least change it to take in account of the pattern. Currently, its just `mon-123`.

Thanks @nizamial09 . Since we have changed the approach and removed the checkbox, Can you please retest this PR?.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -77,30 +89,65 @@ export class HostFormComponent extends CdForm implements OnInit {
 
+  // expand hostnames in case hostname is a pattern
+  checkHostNamePattern(hostname: string) {
+    const braceExpansion = require('brace-expansion');
```
This should be properly imported as we're doing for other packages.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -77,30 +89,65 @@ export class HostFormComponent extends CdForm implements OnInit {
+  // expand hostnames in case hostname is a pattern
+  checkHostNamePattern(hostname: string) {
+    const braceExpansion = require('brace-expansion');
+    if (hostname.includes('[') && hostname.includes(']') && !hostname.match(/(?![^(]*\)),/g)) {
+      const hostnameRange = hostname
```
I find this block perhaps too complicate.

For example, you have already defined a function to detect whether it's a pattern or no not, right? `checkHostNameValue()`... However you cannot use that because that function do too maany things (or at least more than one). That's my suggestion to use simple functions with simple interfaces, and encapsulate them in more complex functions.

This should start like:
```ts

if (isHostnamePattern()) { // you already have that checker
  // then to pattern stuff
} else {
...
}
```

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -77,30 +89,65 @@ export class HostFormComponent extends CdForm implements OnInit {
+    const braceExpansion = require('brace-expansion');
+    if (hostname.includes('[') && hostname.includes(']') && !hostname.match(/(?![^(]*\)),/g)) {
+      const hostnameRange = hostname
+        .replace(/(?<=\d)\s*-\s*(?=\d)/g, '..')
```
Additionally I see lots of regex that are not friendly to interpret:
`/(?![^(]*\)),/g`
`/(?<=\d)\s*-\s*(?=\d)/g`

What about adding a brief explanation nearby?
Additionally, if we are trusting the underlying library (brace-expansion) I'd simply adapt the syntax and let the inner layer do the validation.

So let's check that our charset is there (`,[-]`) and no forbidden chars (`{}`) and perform the[ mapping](https://stackoverflow.com/a/10726800/13173750).

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/host-form/host-form.component.ts:
```diff
@@ -77,30 +89,65 @@ export class HostFormComponent extends CdForm implements OnInit {
+        .replace(/\[/g, '{')
+        .replace(/]/g, '}');
+      this.hostnameArray = braceExpansion(hostnameRange);
+    } else if (hostname.includes(',')) {
+      let hostArray = [];
+      hostArray = hostname.split(',');
+      hostArray.forEach((host: string) => {
+        if (host.includes('[') && host.includes(']')) {
+          const hostnameRange = host
+            .replace(/(?<=\d)\s*-\s*(?=\d)/g, '..')
+            .replace(/\(/g, '[')
+            .replace(/\)/g, ']')
+            .replace(/\[/g, '{')
+            .replace(/]/g, '}');
+          this.hostnameArray = this.hostnameArray.concat(braceExpansion(hostnameRange));
+        } else {
+          this.hostnameArray.push(host);
+        }
+      });
```
BTW we can simply convert the comma-separated list to a brace expression by `{comma-separated-list}`, so perhaps we can simply enclose any input expressions in braces, and remove this else block assimilating it the first block. What do you think?
